### DISCUSSION TITLE
Verbindung RaspberryMatic zu Docker IoBroker mit hm.rpc

### DISCUSSION THREAD

**Discussion by Mephisto090** [OpenCCU/OpenCCU#2567] [🇩🇪 Fragen&Antworten]
Hallo,

ich bekomme irgendwie keine Verbindung mit dem hm-rpc hin, beim hm-rega habe ich irgendwie keine Probleme.

![image](https://github.com/jens-maus/RaspberryMatic/assets/85995176/3dab9abc-0496-4220-9304-5beb9227567f)

Der iobroker im docker läuft auf der Server-IP: `192.168.178.61` (Docker Intern: `172.18.0.2`) und der Raspberrymatic auf der Server IP: `192.168.178.96` (ohne Docker als Raspi-Image)

Portfreigabe ioBroker (docker-compose): 
```
- "1884:1884"
- "8081-8082:8081-8082"
- "8400:8400"
- "42010:42010"
- "2010:2010"
```

Firewall-Raspberrymatic:
![image](https://github.com/jens-maus/RaspberryMatic/assets/85995176/8a2b7cd6-4d24-4fcb-9db6-788bc062d5db)

Serverlog ioBroker hm-rpc:
![image](https://github.com/jens-maus/RaspberryMatic/assets/85995176/ca13c5a8-be28-498c-bcc0-54830fa4d1c5)

Einstellungen RPC:
![image](https://github.com/jens-maus/RaspberryMatic/assets/85995176/50fe8302-1472-4471-aa1a-c845559bb7f9)
![image](https://github.com/jens-maus/RaspberryMatic/assets/85995176/dd98a5c6-9918-43bd-be32-79abcc514150)

Einstellungen Rega:
![image](https://github.com/jens-maus/RaspberryMatic/assets/85995176/196d2878-d9d3-4e9d-b186-fbcda479b404)

Habt ihr eine Lösung oder Idee, bzw. braucht ihr noch irgendwelche Infos?

VG

**Comment by jens-maus** [comment]
Stell mal sicher das du in den hm-rpc.0 Einstellungen bei `Adapter Adresse` auf `IP[v4] 0.0.0.0 - Anfragen für alle IPs beantworten` stellst.

**Reply by Mephisto090** [reply]
![image](https://github.com/jens-maus/RaspberryMatic/assets/85995176/3b636258-01c9-4d50-a3f1-6040e1629574)

habe ich leider nicht

**Reply by Mephisto090** [reply]
@jens-maus hast du noch andere Ideen oder Vorschläge für mich?

**Comment by liquid-void** [comment]
Der Issue ist alt, aber falls jemand immer noch ein ähnliches Problem hat mit: `Unknown XML-RPC tag 'TITLE'`. 

Das ist ein HTML Tag, der angezeigt wird, wenn ein Login popup beim RPC Api Request angezeigt wird. Bedeutet normalerweise, dass in RaspberryMatic die XML-RPC Authentifizierung aktiviert ist, aber der Login nicht richtig funktioniert.
`Startseite > Einstellungen > Systemsteuerung -> Sicherheit -> Authentifizierung aktiv`
Am besten einmal kurz probieren die auszuschalten. Wenn es dann geht, ist das Problem gefunden.

Mit aktivierter Authentifizierung muss beidseitig https aktiviert sein:

- RaspberryMatic > Einstellungen > Systemsteuerung -> Sicherheit -> Authentifizierung aktiv  **und** Automatische Umleitung auf HTTPS aktiv!
![image](https://github.com/user-attachments/assets/a942a101-cecd-4860-82e9-b43b98681ff0)

- hm-rpc Instanz -> Einstellungen -> Zusätzliche Einstellungen -> HTTPS benutzen **und** einen Benutzer mit Passwort, der in RaspberryMatic angelegt wurde
![image](https://github.com/user-attachments/assets/5a1f4670-e40b-4a9f-ac01-7f883b32f9e3)


Zur `Adapter Adresse` in hm-rpc:
`0.0.0.0` sollte ausgewählt sein, wenn du bei einer neuen Instanz keine Änderung vornimmst. Damit funktioniert es bei mir auch mit Docker. 
Natürlich wäre es Quatsch die interne IP des Docker Containers auszuwählen. Warum der anfragende Service überhaupt seine eigene lokale IP konfigurieren muss, verstehe ich nicht. Vielleicht kann mir das jemand erklären? Davon abgesehen frage ich mich, warum man sich nicht für ein freies Eingabefeld zusätzlich zu dem Dropdown entschieden hat. Mit den Browser DevTools kann der Wert wohl auch manipuliert werden.

Update: Mit IP 0.0.0.0 funktioniert zwar eine Verbindung, allerdings kommen längst nicht alle Daten rein und der Gerätemanager in der hm-rpc Konfiguration funkioniert nicht.
Hiermit sollte alles klappen in Docker:  https://github.com/ioBroker/ioBroker.hm-rpc/issues/1345#issuecomment-2632009844

**Reply by Baxxy13** [reply]
> Warum der anfragende Service überhaupt seine eigene lokale IP konfigurieren muss, verstehe ich nicht. Vielleicht kann mir das jemand erklären?

Das ist simpel. 
Der Klient (ioBroker) registriert sich bei der Zentrale am RPC und gibt für den Verbindungsaufbau (Zentrale --> ioBroker) seine Adresse mit. Die Zentrale sendet daraufhin alle Events an die angegebene Adresse.

**Reply by liquid-void** [reply]
Danke für die Info. Da XML-RPC HTTP verwendet, bin ich davon ausgegangen, dass die IP ganz einfach im 'Host' Header des HTTP Requests mitgesendet wird. Ist das also nicht redundant, oder übersehe ich noch etwas?

Beispiel der HTTP Header eines XML-RPC Requests:
https://www.tutorialspoint.com/xml-rpc/xml_rpc_request.htm
```
POST /target HTTP 1.0
User-Agent: Identifier
Host: host.making.request  <-- in diesem fall hm-rpc
Content-Type: text/xml
Content-Length: length of request in bytes
```

Edit:
Achso, da die IP je nach Netwerk-Konfiguration nicht eindeutig ist, muss sie festgelegt werden, damit sie korrekt in den Header geschrieben werden kann, oder?

**Reply by liquid-void** [reply]
Ich bin ehrlich gesagt bisher von einem regelmässigen Polling vom Client zur Zentrale ausgegangen, vor allem da sich Zustände im ioBroker nicht sofort, sondern erst nach ca. 30 Sekunden ändern.
Ist es wirklich so, dass die Zentrale eigenständig POST requests via Subsciption-Modell ausführt? Wie kann das dann funktionieren, wenn man 0.0.0.0 einstellt? Die Zentrale wird ja dann wohl nicht an alle im Netzwerk verfügbaren IPs Requests ausführen.

**Reply by Baxxy13** [reply]
Nein, beim RPC pusht die Zentrale, Polling macht der ReGa-Adapter.
Keine Ahnung was Jens da vor 2 Jahren mit 0.0.0.0 wollte.
Kann man ja nicht auswählen und sollte man auch nicht.
In der einfachen Welt hat der ioBroker eine IP-Adresse aus dem heimischen Netz.
Auf dieser lauscht er und die wird auch an den RPC der Zentrale als Adresse weitergegeben.

Lustig wird's halt nur bei Containern die keine IP-Adresse im heimischen Netz haben.
Hier muss die Callback-Adresse auf die des Container-Hosts gesetzt werden und dieser muss dafür sorgen das die Pakete in den Container gelangen (und umgedreht natürlich). Die Callback-Adresse wird dann als Empfängeradresse beim RPC hinterlegt.

Das Konzept ist hier ganz gut beschrieben und bebildert.
https://flows.nodered.org/node/node-red-contrib-ccu

**Reply by liquid-void** [reply]
@Baxxy13 Danke für den hilfreichen Input. Damit konnte ich jetzt auch erstmal hm-rpc korrekt konfigurieren, sodass alle Daten reinkommen und der Gerätemanager funktioniert:
https://github.com/ioBroker/ioBroker.hm-rpc/issues/1345#issuecomment-2632009844

**Reply by liquid-void** [reply]
Also vorerst nur die simple Lösung mit "host" Netzwerk konfiguriert in Docker.

> Hier muss die Callback-Adresse auf die des Container-Hosts gesetzt werden und dieser muss dafür sorgen das die Pakete in den Container gelangen (und umgedreht natürlich). Die Callback-Adresse wird dann als Empfängeradresse beim RPC hinterlegt.

Das werde ich als nächstes probieren.

Update:
So wie du es erklärt hast, funktioniert es bestens. Man muss nur ein beliebiges Port, z.B. 42035, im Docker Netzwerk durchreichen, welches dann als Adapter Port eingestellt wird. Mit der internen Container IP und der Host IP als Adapter Callback Addresse funktioniert es dann auch mit einem Standard-Dockernetzwerk (Bridge). Danke!
